File: php/509_Fibonacci_Number.php
<?php

class Solution {

     /** 
      * @var $memo 
      */
    public $memo = [
        0 => 0,
        1 => 1,
    ];

    /**
     * @param Integer $n
     * @return Integer
     */
    function fib($n) {
        if (!isset($this->memo[$n])) {
            $this->memo[$n] = $this->fib($n - 1) + $this->fib($n - 2);
        }

        return $this->memo[$n];
    }
}

File: cpp/509_Fibonacci_Number.cpp
class Solution {
    map<int, int> memo = {{0,0}, {1, 1}};
public:
    int fib(int n) {
        if (memo.find(n) == memo.end()) {
            memo[n] = fib(n - 1) + fib(n - 2);
        }
        return memo[n];
    } 
};

